vshn/provider-exoscale
# Add IAM key object

## Summary
This PR adds support for Exoscale IAM Key:
* Adds IAM APIs;
* Adds IAM controller;
* Adds business logic;
* Adds webhook for CRD validation;
* Adds Crossplane package;
* Configures the local environment.

### Insights
* The IAM key in exoscale is immutable which means there are no API available to update a key.
* The key secret is available only once, at creation time. This makes impossible to ensure the API k8s secret's key secret. Thus, if the key secret is lost from the API k8s secret (manually updated/deleted) the IAM key is lost and cannot be used anymore.
* From the reason above, the secret ref cannot be update in the IAM key CRD. Also the key name, zone and allowed bucket list.
* During creation of the IAM key CRD, there should be at least one bucket specified otherwise the webhook will not allow the creation of the IAM key in exoscale. At this point in time our API does not allow to create IAM Keys with unrestricted access to all buckets.

### Next steps
* New PR for docs;
* New PR for webhook tests and e2e kuttle tests.

### How to manually test
1) Access exoscale.com and add a supper IAM key (or at least one that has full access over object storage `sos`). Save the key and secret in env variables `EXOSCALE_API_KEY` and `EXOSCALE_API_SECRET` respectively.
2) If you are using project VSHN AG in exoscale then the bucket `bucket.test` already exists and you don't have to create a new bucket, otherwise a bucket named `bucket.test` has to be created.
3) Run `make local-install install-samples`.
4) Check the IAM key `kubectl get iamkey iam-key`.
5) Delete the IAM key  `kubectl delete iamkey iam-key`.
6) Try to update the IAM key's zone, key name, secret refer or allowed bucket list.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
